### PR TITLE
LAZY LIST LOADING FOR /ATOM/MOVABLE & /TURF/UNSIMULATED/MINERAL

### LINKED ISSUE #13057: snowmap crashed serb
It wasn't in the rotation, which is good, but it still got loaded, which is bad
About twenty minutes in it crashed
This is more a reminder to myself to look into the runtime logs than anything else, also a general reminder to not play snowmap
in terms of bugs, aye there were a few but I think it was a fairly successful round really

Also re: the crash, I'm deeply sorry that happened. I guess it gives you all time to spend christmas with your family or something - Personally, I had the server left on for a day by accident and I didn't get any issues but I can't put into words how apologetic I am this happened even if it was unpreventable, because in the end this wouldn't have happened without snowmap and as soon as the runtime logs are available I'll be working as hard as I can to figure out why

they say a picture means a thousand words so here take this as an apology

### PR DESCRIPTION
**Pull request by ghost**
Will fix #13057 - proving without a doubt this was not a snowmap issue but rather an atom locking & mine turfs issue

Fixes crash from about a week ago that was caused by this.

"Ported" (macros copypasta'd) from https://github.com/tgstation/tgstation/pull/22059 and https://github.com/tgstation/tgstation/pull/21245

### CODE REVIEW

**Comment by PJB3005** [comment]
Yeah fuck off.

If you're gonna touch atom locking, do it properly.

The only reason I haven't fixed atom locking is because legacy (pre-categories) code had to be converted and it was directly accessing the lists.

File: code/__HELPERS/lists.dm [hunk 1]
```diff
@@ -10,6 +10,16 @@
  */
 
 //Returns a list in plain english as a string
```
Comment (ComicIronic) [inline code/__HELPERS/lists.dm]: Move this comment down to where it should be, with the proc.

**Comment by ghost** [comment]
No, you fuck off.
I did a notepad search for the names of the variables changed - none were found other than where I have allowed for this
This is done as properly as It needs to be - if there's a flaw with atom locking code not being adapted properly, I can only blame the original coder

File: code/__HELPERS/lists.dm [hunk 2]
```diff
@@ -10,6 +10,16 @@
...
+
+#define LAZYINITLIST(L) if (!L) L = list()
+#define ISREALLIST(L) (L && istype(L,/list) && L.len)
+#define UNSETEMPTY(L) if(!L.len) L = null
+#define LAZYLEN(L) ( L ? L.len : 0 )
+#define LAZYREMOVE(L, I) if(L) { L -= I; if(!L.len) { L = null; } }
+#define LAZYADD(L, I) if(!L) { L = list(); } L += I;
+#define LAZYACCESS(L, I) (L ? (isnum(I) ? (I > 0 && I <= L.len ? L[I] : null) : L[I]) : null)
+#define LAZYLEN(L) length(L)
```
Comment (ComicIronic) [inline code/__HELPERS/lists.dm]: Bravo on the double define.

Comment (ghost) [reply code/__HELPERS/lists.dm]: thanks

**Comment by PJB3005** [comment]
> This is done as properly as It needs to be

No, not really.

> I can only blame the original coder

I agree, now, how is that ever an excuse to write shitcode?

File: code/__HELPERS/lists.dm (hunk 2 above)
Comment (ghost) [reply code/__HELPERS/lists.dm]: I did say I copypasted the macros from TG, blame them

**Comment by ghost** [comment]
@PJB3005 Yes, really - remove needs fixing

**Review by PJB3005 — CHANGES_REQUESTED** [review]
Do it properly.

File: code/datums/locking_category.dm [hunk 3]
```diff
@@ -81,7 +82,7 @@
 
 	// Okay so now we have to loop through ALL of the owner's locked atoms and their categories to see if the owner still needs to be dense.
 	var/found = FALSE
-	if (flags & DENSE_WHEN_LOCKING || AM.lockflags & DENSE_WHEN_LOCKED)
+	if (flags & DENSE_WHEN_LOCKING || AM.lockflags & DENSE_WHEN_LOCKED && ISREALLIST(owner.locked_atoms))
```
Comment (ComicIronic) [inline code/datums/locking_category.dm]: Should the `||` not be in brackets? `&&` has higher priority.

Comment (ghost) [reply code/datums/locking_category.dm]: Good idea

File: code/game/atoms_movable.dm [hunk 4]
```diff
@@ -61,9 +61,12 @@
 		for(var/matID in starting_materials)
 			materials.addAmount(matID, starting_materials[matID])
 
-	locked_atoms            = list()
-	locking_categories      = list()
-	locking_categories_name = list()
+//  FUCK YOU FUCK YOU FUCK YOU FUCK YOU FUCK YOU
```
Comment (ComicIronic) [inline code/game/atoms_movable.dm]: No meme comments.

**Comment by PJB3005** [comment]
> It didn't even pass travis.

I'm tempted to close this joke of a PR.

File: code/game/objects/structures/crates_lockers/crates.dm [hunk 5]
```diff
@@ -415,7 +415,7 @@
 
 	if(istype(AM, /obj/structure/bed)) //This is only necessary because of rollerbeds and swivel chairs.
 		var/obj/structure/bed/B = AM
-		if(B.locked_atoms.len)
+		if(ISREALLIST(B))
```
Comment (ComicIronic) [inline code/game/objects/structures/crates_lockers/crates.dm]: B.locked_atoms

Comment (ghost) [reply code/game/objects/structures/crates_lockers/crates.dm]: @ComicIronic That'd work but that doesn't check its length or if it's a list. That's why this macro exists

File: code/game/objects/structures/stool_bed_chair_nest/chairs.dm [hunk 6]
```diff
@@ -249,7 +249,7 @@
 
 
 /obj/structure/bed/chair/office/handle_layer() // Fixes layer problem when and office chair is buckled and facing north
-	if(dir == NORTH && !locked_atoms.len)
+	if(dir == NORTH && ISREALLIST(locked_atoms))
```
Comment (ComicIronic) [inline code/game/objects/structures/stool_bed_chair_nest/chairs.dm]: Lost a not here.

**Comment by Probe1** [comment]
Do the needful please.

File: code/game/objects/structures/stool_bed_chair_nest/chairs.dm (hunk 6 above)
Comment (ghost) [reply code/game/objects/structures/stool_bed_chair_nest/chairs.dm]: @ComicIronic Thank you for catching this

File: code/game/objects/structures/crates_lockers/crates.dm (hunk 5 above)
Comment (ghost) [reply code/game/objects/structures/crates_lockers/crates.dm]: Oh wait you mean - yeah I just fixed this

File: code/__HELPERS/lists.dm (hunk 2 above)
Comment (PJB3005) [reply code/__HELPERS/lists.dm]: > blame them.

Don't bring a pig into the house and then blame the owner of the pig when the pig ruins everything.

**Comment by ghost** [comment]
@ComicIronic Addressed your concerns. I'm going to compile and test this works fully now.

@PJB3005 Explain what exactly you mean, please. Maybe I've misinterpret you, but from what I can tell you're asking me to completely refactor atom locking

**Comment by 9600bauds** [comment]
PJB when you make comments like "do it properly" can you please say what you're talking about

**Comment by PJB3005** [comment]
Make a couple damn helpers to check whether a certain atom is locked, etc..., locking category, yada yada.

And then do it properly ffs.

**Comment by TheFriggenSawbones** [comment]
i'd like to thank pjb for allowing me to win the Great Snowmap Bet of 2016

**Comment by ghost** [comment]
So you ARE asking me to completely refactor atom locking code?
I'm still not clear

**Comment by PJB3005** [comment]
Do you even understand how atom locking works in the first place?

**Comment by 9600bauds** [comment]
Can you please answer the question so everyone else knows what you're talking about

**Comment by ghost** [comment]
I'm aware, yes. What issue do you see?

**Comment by PJB3005** [comment]
Ok better idea.

I'll do it. Nerd.

**Comment by 9600bauds** [comment]
PJB I'm fucking begging you for you to use human words and engage in verbal communication

**Comment by PJB3005** [comment]
Ik zou het u toch gewoon compleet moeten uitleggen in het engels, dus het is beter dat ik het zelf doe.

**Comment by ghost** [comment]
Uh, I think that @9600bauds said human, not dutch

**Comment by PJB3005** [comment]
Jammer dan.

Désolé.

**Comment by Senaat1** [comment]
A M S T E R D A M
M
S
T
E
R
D
A
M

**Comment by PJB3005** [comment]
Nee eigenlijk.

België.

**Comment by Senaat1** [comment]
A N T W E R P E N
N
T
W
E
R
P
E
N

**Comment by PJB3005** [comment]
Eigenlijk dichter bij Hasselt maar meer specifiek ga ik niet gaan.

**Comment by Senaat1** [comment]
L U I K
U
I
K

**Comment by Probe1** [comment]
Everyone stop shitposting and fighting please.  It's Christmas (still for me at least) and I ask this one gift from you.  No more fighting today.

**Comment by 9600bauds** [comment]
I've already asked nicely

**Comment by ghost** [comment]
@PJB3005 How long'll it take you to get a PR up with your fix? I'll close this PR when that's merged

**Comment by PJB3005** [comment]
I actually already had the basic helpers and such laid out in a stash but untested and unimplemented.

**Comment by SkowronX** [comment]
lol

**Comment by ghost** [comment]
@PJB3005 So, if this works, what is the problem with merging this and waiting for you to fix atom locking properly? 

That's if it works, I've still got to do some testing

**Comment by ComicIronic** [comment]
Dutchposting is punishable by bike theft.

File: code/game/atoms_movable.dm [hunk 7]
```diff
@@ -85,16 +82,18 @@
...
+	if(ISREALLIST(locked_atoms))
+		for (var/atom/movable/AM in locked_atoms)
+			unlock_atom(AM)
 
 	if (locked_to)
 		locked_to.unlock_atom(src)
 
-	for (var/datum/locking_category/category in locking_categories)
-		qdel(category)
+	if(ISREALLIST(locking_categories))
```
Comment (ghost) [inline code/game/atoms_movable.dm:92]: This segment is tested: It nulls as it should on qdel

File: code/game/atoms_movable.dm [hunk 8]
```diff
@@ -256,6 +255,7 @@
 
 	AM.locked_to = src
 
+	LAZYINITLIST(locked_atoms)
```
Comment (ghost) [inline code/game/atoms_movable.dm:258]: This segment is tested. It correctly creates the locked atoms

File: code/game/atoms_movable.dm [hunk 9]
```diff
@@ -269,6 +269,7 @@
 	locked_atoms    -= AM
 	AM.locked_to     = null
 	category.unlock(AM)
+	UNSETEMPTY(locked_atoms)
```
Comment (ghost) [inline code/game/atoms_movable.dm:272]: This segment is tested, it correctly nulls the locked atoms list when there's nothing in it anymore

File: code/game/machinery/doors/door.dm [hunk 10]
```diff
@@ -89,7 +89,7 @@ var/list/all_doors = list()
 		var/obj/structure/bed/chair/vehicle/vehicle = AM
 
 		if (density)
-			if (vehicle.locked_atoms.len && !operating && allowed(vehicle.locked_atoms[1]))
+			if (vehicle.locked_atoms.len && !operating && ISREALLIST(vehicle.locked_atoms) && allowed(vehicle.locked_atoms[1]))
```
Comment (ghost) [inline code/game/machinery/doors/door.dm:92]: This line is tested - the code correctly determines if the door access if the vehicle occupant has access, and doesn't runtime when an empty vehicle is thrown in via buildmode to the door

File: code/game/objects/structures/cage.dm [hunk 11]
```diff
@@ -223,7 +223,7 @@
 			to_chat(victim, "<span class='notice'>You suddenly find yourself locked in a cage!</span>")
 
 /obj/structure/cage/proc/mob_is_inside(mob/checked)
-	return ((contents.Find(checked)) || (locked_atoms.Find(checked)))
+	return ((contents.Find(checked)) || (ISREALLIST(locked_atoms) && locked_atoms.Find(checked)))
```
Comment (ghost) [inline code/game/objects/structures/cage.dm:226]: This code works correctly - it returns 1 if there's someone locked to it and null if there isn't. No runtimes.

File: code/game/objects/structures/crates_lockers/crates.dm [hunk 12]
```diff
@@ -415,7 +415,7 @@
 
 	if(istype(AM, /obj/structure/bed)) //This is only necessary because of rollerbeds and swivel chairs.
 		var/obj/structure/bed/B = AM
-		if(B.locked_atoms.len)
+		if(ISREALLIST(B.locked_atoms))
```
Comment (ghost) [inline code/game/objects/structures/crates_lockers/crates.dm:418]: This code is tested. Crates will take in rollerbeds and swivel chairs if they have no occupants but will not if they do; no runtimes.

File: code/game/objects/structures/electricchair.dm [hunk 13]
```diff
@@ -63,7 +63,7 @@
 		return
 	if(!A.powered(EQUIP))
 		return
-	if(locked_atoms.len)
+	if(ISREALLIST(locked_atoms))
```
Comment (ghost) [inline code/game/objects/structures/electricchair.dm:66]: This works as intended - shocks the human if he's buckled, otherwise just runs the
``` 
spark_system.start() //just something to let them know it works``` in the else

**Comment by SysadminJeroen** [comment]
@DrCelt Wat zeg jij? Duyvelspecht! Impliceer jij dat wij Nederlandstalige geen mensen zijn?

@ComicIronic Dat is niet zo'n probleem als je er achter komt dat ik mijn schroot van een fiets juist kwijt wil raken en al 4 á 5 maanden in de voortuin staat.

File: code/game/objects/structures/popout_cake.dm [hunk 14]
```diff
@@ -43,7 +43,10 @@
 	var/string_pulled = 0
 
 /obj/structure/popout_cake/Destroy()
-	for(var/mob/living/L in locked_atoms + contents) //Release all mobs inside
```
Comment (ghost) [inline code/game/objects/structures/popout_cake.dm:46]: This change doesn't work, but I recently learned for x in null is fine syntax so I reckon we can revert to the old code

File: code/game/objects/structures/popout_cake.dm [hunk 15]
```diff
@@ -138,7 +141,7 @@
 	if(!istype(L))
 		return
 
-	if(locked_atoms.Find(L))
+	if(ISREALLIST(locked_atoms) && locked_atoms.Find(L))
```
Comment (ghost) [inline code/game/objects/structures/popout_cake.dm:142]: Works. If you are popping out of the cake you can still walk out of the cake

File: code/game/objects/structures/stool_bed_chair_nest/alien_nests.dm [hunk 16]
```diff
@@ -39,7 +39,7 @@
 		src.add_fingerprint(user)
 
 /obj/structure/bed/nest/buckle_mob(mob/M as mob, mob/user as mob)
-	if (locked_atoms.len || !ismob(M) || (get_dist(src, user) > 1) || (M.loc != src.loc) || user.restrained() || user.stat || M.locked_to || istype(user, /mob/living/silicon/pai) )
+	if (ISREALLIST(locked_atoms) || !ismob(M) || (get_dist(src, user) > 1) || (M.loc != src.loc) || user.restrained() || user.stat || M.locked_to || istype(user, /mob/living/silicon/pai) )
```
Comment (ghost) [inline code/game/objects/structures/stool_bed_chair_nest/alien_nests.dm:42]: This code works

File: code/game/objects/structures/stool_bed_chair_nest/alien_nests.dm [hunk 17]
```diff
@@ -90,7 +90,7 @@
 		qdel(src)
 
 /obj/structure/bed/nest/proc/stabilize()
-	if(!locked_atoms || !locked_atoms.len)
+	if(!ISREALLIST(locked_atoms))
```
Comment (ghost) [inline code/game/objects/structures/stool_bed_chair_nest/alien_nests.dm:93]: This code works, no runtimes when calling stabilize with or without occupant

File: code/game/objects/structures/stool_bed_chair_nest/alien_nests.dm [hunk 18]
```diff
@@ -101,7 +101,7 @@
 		return
 
 	spawn(15)
-		if(!gcDestroyed && locked_atoms.len)
+		if(!gcDestroyed && ISREALLIST(locked_atoms))
```
Comment (ghost) [inline code/game/objects/structures/stool_bed_chair_nest/alien_nests.dm:104]: This code works. No runtimes if an occupant leaves the nest

File: code/game/objects/structures/stool_bed_chair_nest/bed.dm [hunk 19]
```diff
@@ -52,7 +52,7 @@
 	buckle_mob(M, user)
 
 /obj/structure/bed/proc/manual_unbuckle(mob/user as mob)
-	if(!locked_atoms.len)
+	if(!ISREALLIST(locked_atoms))
```
Comment (ghost) [inline code/game/objects/structures/stool_bed_chair_nest/bed.dm:55]: This code works. You can unbuckle from beds and all their children

File: code/game/objects/structures/stool_bed_chair_nest/bed.dm [hunk 20]
```diff
@@ -162,7 +162,7 @@
 		if(!ishuman(usr) || usr.incapacitated() || usr.lying)
 			return
 
-		if(locked_atoms.len)
+		if(ISREALLIST(locked_atoms))
```
Comment (ghost) [inline code/game/objects/structures/stool_bed_chair_nest/bed.dm:165]: This code works. You can't collapse a rollerbed someone is on

File: code/game/objects/structures/stool_bed_chair_nest/chairs.dm [hunk 21]
```diff
@@ -166,7 +166,7 @@
 
 /obj/structure/bed/chair/comfy/update_icon()
 	..()
-	if(locked_atoms.len)
+	if(ISREALLIST(locked_atoms))
```
Comment (ghost) [inline code/game/objects/structures/stool_bed_chair_nest/chairs.dm:169]: This code works. The overlays appear correctly when unlocking/locking the human

File: code/game/objects/structures/stool_bed_chair_nest/chairs.dm [hunk 22]
```diff
@@ -195,7 +195,7 @@
 	return ..()
 
 /obj/structure/bed/chair/comfy/attack_hand(var/mob/user)
-	if(locked_atoms.len)
+	if(ISREALLIST(locked_atoms))
```
Comment (ghost) [inline code/game/objects/structures/stool_bed_chair_nest/chairs.dm:198]: This code works. It returns ..() if there is an occupant and doesn't if there isn't.

File: code/game/objects/structures/stool_bed_chair_nest/chairs.dm [hunk 23]
```diff
@@ -240,7 +240,7 @@
 
 /obj/structure/bed/chair/office/update_icon()
 	..()
-	if(locked_atoms.len)
+	if(ISREALLIST(locked_atoms))
```
Comment (ghost) [inline code/game/objects/structures/stool_bed_chair_nest/chairs.dm:243]: This code works, the buckle overlays if there is a mob and not if there isn't

File: code/game/objects/structures/stool_bed_chair_nest/chairs.dm [hunk 24]
```diff
@@ -249,7 +249,7 @@
 
 
 /obj/structure/bed/chair/office/handle_layer() // Fixes layer problem when and office chair is buckled and facing north
-	if(dir == NORTH && !locked_atoms.len)
+	if(dir == NORTH && !ISREALLIST(locked_atoms))
```
Comment (ghost) [inline code/game/objects/structures/stool_bed_chair_nest/chairs.dm:252]: Works

File: code/game/objects/structures/vehicles/adminbus_powers.dm [hunk 25]
```diff
@@ -331,7 +331,7 @@
 
 	var/distributed = 0
 
-	if(locked_atoms.len)
+	if(ISREALLIST(locked_atoms))
```
Comment (ghost) [inline code/game/objects/structures/vehicles/adminbus_powers.dm:334]: I hate adminbus code but this works fine

File: code/game/objects/structures/vehicles/adminbus_powers.dm [hunk 26]
```diff
@@ -393,7 +393,7 @@
...
```
Comment (ghost) [inline code/game/objects/structures/vehicles/adminbus_powers.dm:396]: Works

File: code/game/objects/structures/vehicles/wheelchair.dm [hunk 27]
```diff
@@ -169,7 +169,7 @@
 /obj/structure/bed/chair/vehicle/wheelchair/multi_people/examine(mob/user)
 	..()
 
-	if(locked_atoms.len > 9)
+	if(ISREALLIST(locked_atoms) && locked_atoms.len > 9)
```
Comment (ghost) [inline code/game/objects/structures/vehicles/wheelchair.dm:172]: why does this item exist. Yes, no runtimes, functionality still there

File: code/game/objects/structures/vehicles/wheelchair.dm [hunk 28]
```diff
@@ -180,6 +180,8 @@
 
 /obj/structure/bed/chair/vehicle/wheelchair/multi_people/update_mob()
 	var/i = 0
+	if(!ISREALLIST(locked_atoms))
```
Comment (ghost) [inline code/game/objects/structures/vehicles/wheelchair.dm]: This code is unnecessary, x in null doesn't run, I will remove this

File: code/game/supplyshuttle.dm [hunk 29]
```diff
@@ -70,7 +70,7 @@ var/list/mechtoys = list(
 		return prob(60)
 
 	var/obj/structure/bed/B = mover
-	if (istype(mover, /obj/structure/bed) && B.locked_atoms.len)//if it's a bed/chair and someone is locked_to, it will not pass
+	if (istype(mover, /obj/structure/bed) && ISREALLIST(B.locked_atoms) && B.locked_atoms.len)//if it's a bed/chair and someone is locked_to, it will not pass
```
Comment (ghost) [inline code/game/supplyshuttle.dm:73]: Works - you can't push a human through plasticflaps on a wheelie chair

File: code/modules/events/immovablerod.dm [hunk 30]
```diff
@@ -49,7 +49,6 @@
 	throwforce = 100
 	density = 1
 	anchored = 1
-	locked_atoms = list()
```
Comment (ghost) [inline code/modules/events/immovablerod.dm:52]: Doesn't need to be here. Did get a runtime from spawning an immovable rod. Since fixed

File: code/modules/hydroponics/spreading/spreading_growth.dm [hunk 31]
```diff
@@ -63,7 +63,7 @@
 	update_icon()
 
 	if(is_mature() && special_cooldown())
-		if(locked_atoms && locked_atoms.len)
+		if(ISREALLIST(locked_atoms))
```
Comment (ghost) [inline code/modules/hydroponics/spreading/spreading_growth.dm:66]: works

File: code/modules/hydroponics/spreading/spreading_response.dm [hunk 32]
```diff
@@ -113,7 +113,7 @@
 
 /obj/effect/plantsegment/proc/manual_unbuckle(mob/user as mob)
 	var/list/atom/movable/locked = get_locked(/datum/locking_category/plantsegment)
-	if(locked && locked.len)
+	if(ISREALLIST(locked))
```
Comment (ghost) [inline code/modules/hydroponics/spreading/spreading_response.dm:116]: works

File: code/modules/hydroponics/spreading/vines.dm [hunk 33]
```diff
@@ -139,7 +139,7 @@
 
 	if(prob(seed ? seed.potency : 25))
 		var/list/atom/movable/locked = get_locked(/datum/locking_category/plantsegment)
-		if(!locked.len)
+		if(ISREALLIST(locked))
```
Comment (ghost) [inline code/modules/hydroponics/spreading/vines.dm:142]: works

File: code/modules/mining/mine_turfs.dm [hunk 34]
```diff
@@ -18,7 +18,7 @@
 	var/last_act = 0
 	var/datum/geosample/geologic_data
 	var/excavation_level = 0
-	var/list/finds = list()//no longer null to prevent those pesky runtime errors
```
Comment (ghost) [inline code/modules/mining/mine_turfs.dm:21]: I can just imagine an oldcoder going "eh well we COULD use sanity but why not just create a new list for every single mineral turf that won't cause problems right?"

File: code/modules/mining/mine_turfs.dm [hunk 35]
```diff
@@ -263,7 +182,7 @@ turf/unsimulated/mineral/ChangeTurf(var/turf/N, var/tell_universe=1, var/force_l
 
 		var/fail_message = ""
 		//handle any archaeological finds we might uncover
-		if(finds && finds.len)
+		if(ISREALLIST(finds))
```
Comment (ghost) [inline code/modules/mining/mine_turfs.dm:185]: works

File: code/modules/mining/mine_turfs.dm [hunk 36]
```diff
@@ -275,6 +194,7 @@ turf/unsimulated/mineral/ChangeTurf(var/turf/N, var/tell_universe=1, var/force_l
 				excavate_find(5, finds[1])
 			else if(prob(50))
 				finds.Remove(finds[1])
+				UNSETEMPTY(finds)
```
Comment (ghost) [inline code/modules/mining/mine_turfs.dm:197]: works

File: code/modules/mining/mine_turfs.dm [hunk 37]
```diff
@@ -283,7 +203,7 @@ turf/unsimulated/mineral/ChangeTurf(var/turf/N, var/tell_universe=1, var/force_l
 		if(do_after(user, src, P.digspeed) && user)
 			busy = 0
 
-			if(finds && finds.len)
+			if(ISREALLIST(finds))
```
Comment (ghost) [inline code/modules/mining/mine_turfs.dm:206]: works

File: code/modules/mining/mine_turfs.dm [hunk 38]
```diff
@@ -332,7 +252,7 @@ turf/unsimulated/mineral/ChangeTurf(var/turf/N, var/tell_universe=1, var/force_l
 								polarstar = 2
 				return
 
-			if(finds && finds.len)
+			if(ISREALLIST(finds))
```
Comment (ghost) [inline code/modules/mining/mine_turfs.dm:255]: works

File: code/modules/mining/mine_turfs.dm [hunk 39]
```diff
@@ -347,7 +267,7 @@ turf/unsimulated/mineral/ChangeTurf(var/turf/N, var/tell_universe=1, var/force_l
 
 			excavation_level += P.excavation_amount
 
-			if(!archaeo_overlay && finds && finds.len)
+			if(!archaeo_overlay && ISREALLIST(finds))
```
Comment (ghost) [inline code/modules/mining/mine_turfs.dm:270]: works

File: code/modules/mining/mine_turfs.dm [hunk 40]
```diff
@@ -458,6 +378,7 @@ turf/unsimulated/mineral/ChangeTurf(var/turf/N, var/tell_universe=1, var/force_l
 				X = null
 
 	finds.Remove(F)
+	UNSETEMPTY(finds)
```
Comment (ghost) [inline code/modules/mining/mine_turfs.dm:381]: works

File: code/modules/mining/mine_turfs.dm [hunk 41]
```diff
@@ -627,7 +548,10 @@ turf/unsimulated/mineral/ChangeTurf(var/turf/N, var/tell_universe=1, var/force_l
 
 /turf/unsimulated/mineral/random
 	name = "Mineral deposit"
-	var/mineralSpawnChanceList = list(
+	var/mineralChance = 10  //means 10% chance of this plot changing to a mineral deposit
+
+/turf/unsimulated/mineral/random/proc/return_mineral_spawn_chance_list()
```
Comment (ghost) [inline code/modules/mining/mine_turfs.dm:553]: Works, spawned a bunch of random minerals and it was fine

File: code/modules/mining/mine_turfs.dm [hunk 42]
```diff
@@ -675,54 +582,27 @@ turf/unsimulated/mineral/ChangeTurf(var/turf/N, var/tell_universe=1, var/force_l
 /turf/unsimulated/mineral/random/high_chance
 	icon_state = "rock(high)"
 	mineralChance = 25
-	mineralSpawnChanceList = list(
+
+/turf/unsimulated/mineral/random/high_chance/return_mineral_spawn_chance_list()
```
Comment (ghost) [inline code/modules/mining/mine_turfs.dm:586]: Also spawned this - worked fine

File: code/modules/mining/mine_turfs.dm [hunk 43]
```diff
@@ -675,54 +582,27 @@ turf/unsimulated/mineral/ChangeTurf(var/turf/N, var/tell_universe=1, var/force_l
...
-		"Cytine"  = 5
-		*/
 	)
 
 /turf/unsimulated/mineral/random/high_chance_clown
 	icon_state = "rock(clown)"
 	mineralChance = 40
-	mineralSpawnChanceList = list(
+
+/turf/unsimulated/mineral/random/high_chance_clown/return_mineral_spawn_chance_list()
```
Comment (ghost) [inline code/modules/mining/mine_turfs.dm:600]: Same here

File: code/modules/mob/living/simple_animal/hostile/alien.dm [hunk 44]
```diff
@@ -144,7 +144,7 @@ var/list/nest_locations = list()
 	var/smallest_distance = 555
 	var/best_nest
 	for(var/obj/structure/bed/nest/N in nest_locations)
-		if((N.z == src.z) && (N.locked_atoms.len == 0))
+		if((N.z == src.z) && (!ISREALLIST(N.locked_atoms)))
```
Comment (ghost) [inline code/modules/mob/living/simple_animal/hostile/alien.dm:147]: I'm not sure how the old line passed peer review but whatever

**Comment by Senaat1** [comment]
Wilhelmus van Nassouwe
ben ik, van Duitsen bloed,
den Vaderland getrouwe
blijf ik tot in den dood.
Een prinse van Oranje
ben ik, vrij onverveerd,
den Koning van Hispanje
heb ik altijd geëerd.

File: code/modules/mob/living/simple_animal/hostile/mimic.dm [hunk 45]
```diff
@@ -280,7 +280,7 @@ var/global/list/crate_mimic_disguises = list(\
 	if(can_grab && istype(target, /mob/living/carbon/human))
 		var/mob/living/carbon/human/H = target
 		var/list/atom/movable/locked = get_locked(/datum/locking_category/mimic)
-		if(!locked.len) //Eating nobody
+		if(ISREALLIST(locked)) //Eating nobody
```
Comment (ghost) [inline code/modules/mob/living/simple_animal/hostile/mimic.dm:283]: works

File: code/modules/mob/living/simple_animal/hostile/mimic.dm [hunk 46]
```diff
@@ -322,7 +322,7 @@ var/global/list/crate_mimic_disguises = list(\
 
 /mob/living/simple_animal/hostile/mimic/crate/chest/attackby(obj/item/W, mob/user)
 	var/list/atom/movable/locked = get_locked(/datum/locking_category/mimic)
-	if (angry && locked.len && W.is_sharp())
+	if (angry && ISREALLIST(locked) && W.is_sharp())
```
Comment (ghost) [inline code/modules/mob/living/simple_animal/hostile/mimic.dm:325]: works

File: code/modules/research/xenoarchaeology/misc.dm [hunk 47]
```diff
@@ -15,7 +15,7 @@ proc/SetupXenoarch()
 			var/turf/unsimulated/mineral/archeo_turf = turfs_to_process[1]
 
 			for(var/turf/unsimulated/mineral/T in orange(1, archeo_turf))
-				if(T.finds.len)
+				if(ISREALLIST(T.finds))
```
Comment (ghost) [inline code/modules/research/xenoarchaeology/misc.dm:18]: works as far as I can tell

File: code/modules/research/xenoarchaeology/misc.dm [hunk 48]
```diff
@@ -27,8 +27,8 @@ proc/SetupXenoarch()
 			turfs_to_process.Remove(archeo_turf)
 			processed_turfs.Add(archeo_turf)
 
-			if(!archeo_turf.finds || !archeo_turf.finds.len)
-
+			if(!ISREALLIST(archeo_turf.finds))
```
Comment (ghost) [inline code/modules/research/xenoarchaeology/misc.dm:30]: yeah it works

**Review by ghost — COMMENTED** [review]
:100: tested

>can't approve your own PR
WOW

**Comment by TheFriggenSawbones** [comment]
Ik haat mijn leven

**Comment by ghost** [comment]
Alright I squashed it into one commit so I can cherrypick it to test if snowmap works without crashing now. I'd like some peer review - it does work. It is tested, I don't see what @PJB3005 is talking about and I'd love him to elaborate.

In my opinion this is absolutely ready to be merged but I still wanna hear @PJB3005's reasons for requesting changes

**Comment by 9600bauds** [comment]
some way or another there's gonna be a lot of effort wasted over complete lack of communication here

**Comment by Kitfox88** [comment]
i'm gay

**Review by PJB3005 — CHANGES_REQUESTED** [review]
Not happening

**Comment by Probe1** [comment]
Please explain why.  You've only said that it should not be merged and that you will do it.

**Comment by PJB3005** [comment]
Because it's a complete half-assed bandaid instead of doing it properly?

**Comment by Probe1** [comment]
You have to explain what that means even if you're frustrated with Celt.

Step out what needs to be done for your approval.

**Comment by PJB3005** [comment]
I have, and Celt said he can't be bothered to do it properly.

**Comment by Probe1** [comment]
I just read the damn PR and there's nothing I can point to that thoroughly explains the changes you're requesting.  

And I did translate all the comments.

**Comment by 9600bauds** [comment]
PJB, listen to me very carefully.

I need you to stop being an insidious little fuck and drop the "intentionally obtuse" act.
Then I need you to type one complete human sentence explaining your alleged review here.
Until you can do that I see nothing wrong with this PR.

Every single person that pours their effort into this repository is here because they want to contribute to the game, nothing more, nothing less, and every single person in this repository could bear you acting a little bit more collaborative already.

**Comment by PJB3005** [comment]
Well good thing you're all bad at encapsulation then, just like I was when I initially wrote atom locking.

File: code/__HELPERS/lists.dm (hunk 2 above)
Comment (RemieRichards) [reply code/__HELPERS/lists.dm]: Actually, not even our fault, if you look at the commits we removed the first one FOR the second one (```length()``` call is actually cheaper than the ternary, cos byond), so we never had both at once.

**Comment by ghost** [comment]
reopen this please I did comment in his PR not to autoclose this but I guess he didn't see it

**Comment by PJB3005** [comment]
I will still shoot you if you even dare to use those readability ruining macros, though.

**Comment by Probe1** [comment]
We're all going to work together.

Or else.

**Comment by 9600bauds** [comment]
Why wouldn't you just open a new PR for this?

**Comment by Probe1** [comment]
It'd really be for the best at this point especially after you wrote comments on about 30 lines.

**Comment by ghost** [comment]
Sure, I'll open a new PR
